src/auspex/stream.py

```diff
@@ -222,14 +266,22 @@ def last_data_axis(self):
             logger.warning("DataStreamDescriptor has no pure DataAxis. Return None.")
             return None
 
-    def axis_data_type(self, with_metadata=False):
+    def axis_data_type(self, with_metadata=False, excluding_axis=None):
         dtype = []
         for a in self.axes:
-            dtype.extend(a.data_type(with_metadata=with_metadata))
-        # dtype.append((self.data_name, self.dtype))
+            if a.name != excluding_axis:
+                dtype.extend(a.data_type(with_metadata=with_metadata))
         return dtype
 
-    def tuples(self, with_metadata=False, as_structured_array=True):
+    def tuples(self, as_structured_array=True):
+        if as_structured_array:
+            # If we already have a structured array
+            if type(self.visited_tuples) is np.ndarray and type(self.visited_tuples.dtype.names) is tuple:
+                return self.visited_tuples
+            return np.core.records.fromrecords(self.visited_tuples, dtype=self.axis_data_type(with_metadata=True))
+        return self.visited_tuples
+
+    def expected_tuples(self, with_metadata=False, as_structured_array=True):
         vals           = [a.points_with_metadata() for a in self.axes]
         nested_list    = list(itertools.product(*vals))
```

Comment:
What's the point of calling `list` on this if you are going to iterate over it on the next line?